### PR TITLE
Brig door timer tgui

### PR DESCRIPTION
**Pull request by AnturK**

### CODE REVIEW

File: code/game/machinery/doors/brigdoors.dm [hunk 1]
```diff
@@ -259,6 +192,54 @@
...
+
+/obj/machinery/door_timer/ui_act(action, params)
+	if(..())
+		return
+
+	switch(action)
+		if("time")
+			var/value = text2num(params["mod"])
+			set_timer(time_left()+value)
+		if("start")
```
Comment (neersighted) [inline code/game/machinery/doors/brigdoors.dm]: remember to `. = TRUE` if an action caused a UI update

File: code/game/machinery/doors/brigdoors.dm [hunk 2]
```diff
@@ -259,6 +192,54 @@
...
+		if(F.last_flash && (F.last_flash + 150) > world.time)
+			data["flash_charging"] = TRUE
+			break
+	return data
+
+
+/obj/machinery/door_timer/ui_act(action, params)
+	if(..())
+		return
+
```
Comment (neersighted) [inline code/game/machinery/doors/brigdoors.dm]: no space here

File: tgui/src/interfaces/brig_timer.ract [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    <span>{{data.minutes}}:{{data.seconds}}</span>
+    <ui-button icon='forward' action='time' params='{"mod": 100}'></ui-button>
+    <ui-button icon='fast-forward' action='time' params='{"mod": 600}'></ui-button>
+  </ui-section>
+  <ui-section>
+      <ui-button icon='hourglass-start' action='preset' params='{"preset": "short"}'>Short</ui-button>
+      <ui-button icon='hourglass-start' action='preset' params='{"preset": "medium"}'>Medium</ui-button>
+      <ui-button icon='hourglass-start' action='preset' params='{"preset": "long"}'>Long</ui-button>
+  </ui-section>
+  {{#partial button}}
```
Comment (neersighted) [inline tgui/src/interfaces/brig_timer.ract]: Partial at the top please

File: tgui/src/interfaces/brig_timer.ract [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    {{#if data.timing}}
+          <ui-button icon='clock-o' style='selected' action='stop'>Stop</ui-button>
```
Comment (neersighted) [inline tgui/src/interfaces/brig_timer.ract]: Make these ternaries; so we don't create/destroy dom elements (slow).

File: tgui/src/interfaces/brig_timer.ract [hunk 5]
```diff
@@ -0,0 +1,25 @@
+<script>
+  component.exports = {
+    data: {
+      text: require('util/text')
```
Comment (neersighted) [inline tgui/src/interfaces/brig_timer.ract]: this is not needed, text is defined on the base component.

File: tgui/src/interfaces/brig_timer.ract [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    }
+  }
+</script>
+<ui-display title='Cell Timer' button>
+  {{#partial button}}
+    <ui-button icon='clock-o' style='{{data.timing ? "selected" : null}}' action='{{data.timing ? "stop" : "start"}}'>{{data.timing ? "Stop" : "Start"}}</ui-button>
+    <ui-button icon='lightbulb-o' action='flash' style='{{data.flash_charging ? "disabled" : null}}'>{{data.flash_charging ? "Recharging" : "Flash"}}</ui-button>
+  {{/partial}}
+  <ui-section>
+    <ui-button icon='fast-backward' action='time' params='{"mod": -600}'></ui-button>
```
Comment (neersighted) [inline tgui/src/interfaces/brig_timer.ract]: I'd name this parameter adjust for consistency.

File: tgui/src/util/text.js [hunk 7]
```diff
@@ -5,3 +5,10 @@ export function upperCaseFirst (str) {
 export function titleCase (str) {
   return str.replace(/\w\S*/g, upperCaseFirst)
 }
+
+export function zeroPad (str, pad_size) {
+  str = str.toString()
+  while(str.length < pad_size)
+  	str = "0" + str
```
Comment (neersighted) [inline tgui/src/util/text.js]: Single quotes for string literals.

File: code/game/machinery/doors/brigdoors.dm [hunk 8]
```diff
@@ -38,24 +44,22 @@
...
+	for(var/obj/structure/closet/secure_closet/brig/C in ultra_range(20, src))
+		if(C.id == id)
+			targets += C
 
-		if(targets.len==0)
-			stat |= BROKEN
-		update_icon()
-		return
-	return
+	if(targets.len==0)
```
Comment (neersighted) [inline code/game/machinery/doors/brigdoors.dm]: `!targets.len`

File: code/game/machinery/doors/brigdoors.dm [hunk 9]
```diff
@@ -259,6 +192,57 @@
...
+	return data
+
+
+/obj/machinery/door_timer/ui_act(action, params)
+	if(..())
+		return
+	. = TRUE
+	switch(action)
+		if("time")
+			var/value = text2num(params["adjust"])
```
Comment (neersighted) [inline code/game/machinery/doors/brigdoors.dm]: What would happen here if `value` was null?

File: tgui/src/interfaces/brig_timer.ract [hunk 10]
```diff
@@ -0,0 +1,18 @@
+<ui-display title='Cell Timer' button>
+  {{#partial button}}
+    <ui-button icon='clock-o' style='{{data.timing ? "selected" : null}}' action='{{data.timing ? "stop" : "start"}}'>{{data.timing ? "Stop" : "Start"}}</ui-button>
+    <ui-button icon='lightbulb-o' action='flash' style='{{data.flash_charging ? "disabled" : null}}'>{{data.flash_charging ? "Recharging" : "Flash"}}</ui-button>
+  {{/partial}}
+  <ui-section>
+    <ui-button icon='fast-backward' action='time' params='{"adjust": -600}'></ui-button>
+    <ui-button icon='backward' action='time' params='{"adjust": -100}'></ui-button>
+    <span>{{text.zeroPad(data.minutes,2)}}:{{text.zeroPad(data.seconds,2)}}</span>
```
Comment (neersighted) [inline tgui/src/interfaces/brig_timer.ract]: Space out function calls please

**Comment by neersighted** [comment]
All that remains are minor style quibbles. Looks good, :+1:.

**Comment by freerealestate** [comment]
should have hardcoded sentences for specific crimes to prevent abuse tbh

**Comment by neersighted** [comment]
:+1: mergeit

**Pull request merged**